### DISCUSSION TITLE
Comparing pure floating point computation between bun,node,qjs,deno,jsc

### DISCUSSION THREAD

**Discussion by mingodad** [oven-sh/bun#1719] [General]
### What version of Bun is running?

0.4.0

### What platform is your computer?

Linux 5.4.0-135-generic x86_64 x86_64

### What steps can reproduce the bug?

Using tests from https://github.com/ArashPartow/exprtk adapted to Javascript (see attached 7445 expressions) and testing it with bun, node, quickjs it's interesting that quickjs is faster than both node and bun in the given order, and memory usage is quickjs 6.3MB, node 46.2MB and bun 98.9MB.

```
/usr/bin/time qjs "exprtk_functional_test.txt.js"
7445 7299
0.07user 0.00system 0:00.07elapsed 98%CPU (0avgtext+0avgdata 6300maxresident)k
0inputs+0outputs (0major+1714minor)pagefaults 0swaps

/usr/bin/time node-env node "exprtk_functional_test.txt.js"
7445 7299
0.08user 0.01system 0:00.10elapsed 102%CPU (0avgtext+0avgdata 46204maxresident)k
0inputs+0outputs (0major+6505minor)pagefaults 0swaps

/usr/bin/time bun "exprtk_functional_test.txt.js"
7445 7299
0.11user 0.05system 0:00.16elapsed 100%CPU (0avgtext+0avgdata 98980maxresident)k
0inputs+0outputs (0major+17494minor)pagefaults 0swaps

/usr/bin/time bun "exprtk_functional_test.txt.ts"
7445 7299
0.14user 0.02system 0:00.17elapsed 100%CPU (0avgtext+0avgdata 100488maxresident)k
0inputs+0outputs (0major+17769minor)pagefaults 0swaps
```

[exprtk_functional_test.txt.js.zip](https://github.com/oven-sh/bun/files/10344809/exprtk_functional_test.txt.js.zip)

**Comment by mingodad** [comment]
Also here is the output of deno 1.29.1:
```
/usr/bin/time deno run "exprtk_functional_test.txt.js"
7445 7299
0.06user 0.02system 0:00.09elapsed 102%CPU (0avgtext+0avgdata 51244maxresident)k
0inputs+64outputs (0major+7171minor)pagefaults 0swaps
```

**Reply by guest271314** [reply]
QuickJS is faster than Deno and Node.js as a Native Messaging host, too https://gist.github.com/guest271314/fb8969a956a873e495b890a42ac0d731.

I just finally created a working Bun Native Messaging host. I'll include Bun in the test and post here.

Not surprising to me. With `strip` applied `qjs` is less than 1MB. That is one reason QuickJS is favored as a JavaScript runtime in WASM applications, e.g., https://github.com/second-state/wasmedge-quickjs.

**Reply by guest271314** [reply]
I updated the aforementioned Native Messaging host test, creating 100 Native Messaging hosts for each JavaScript engine and echoing an array of length `10000`. Result (average time to complete):

```
for (const [key, prop] of Object.entries(temp1)) {
  console.log(key, prop.reduce((a, b) => a += b, 0) / prop.length);
}
// nm_qjs 0.09441000000003727
// nm_deno 0.23672499999992547
// nm_nodejs 0.3944149999999626
// nm_bun 0.17533899999976157
```

Not surprising QuickJS is the fastest. Bun 2d fastest. Deno 3d fastest. Node.js slowest.

**Comment by Jarred-Sumner** [comment]
converted to discussion because it’s not a perf problem in an app or library

**Comment by mingodad** [comment]
After building this project https://github.com/mean-ui-thread/JavaScriptCore and testing I'm getting this output:
```
/usr/bin/time myjsc exprtk_functional_test.txt.js 
--> 7445 7299
0.08user 0.03system 0:00.12elapsed 100%CPU (0avgtext+0avgdata 63420maxresident)k
0inputs+0outputs (0major+11893minor)pagefaults 0swaps
```
Taking in consideration that `bun` also uses `JavascriptCore` then it seems that something in the `bun` build is making it slower and  memory hungry for this particular test.

**Reply by mingodad** [reply]
For the test above I needed to add this line to the test source:
```
console = console || { log: (...args) => debug(Array.prototype.slice.call(args).join(' ')) }
```

**Comment by mingodad** [comment]
Interesting testing `deno` it takes a bit more time the first cold run than `bun`, `jsc` and `node`:
```
/usr/bin/time deno run exprtk_functional_test.txt.js 
7445 7299
0.24user 0.03system 0:00.28elapsed 97%CPU (0avgtext+0avgdata 81440maxresident)k
0inputs+112outputs (0major+13245minor)pagefaults 0swaps

$ /usr/bin/time deno run exprtk_functional_test.txt.js 
7445 7299
0.07user 0.01system 0:00.08elapsed 101%CPU (0avgtext+0avgdata 51116maxresident)k
0inputs+64outputs (0major+7184minor)pagefaults 0swaps

$ /usr/bin/time deno run exprtk_functional_test.txt.js 
7445 7299
0.07user 0.01system 0:00.09elapsed 101%CPU (0avgtext+0avgdata 50272maxresident)k
0inputs+64outputs (0major+7175minor)pagefaults 0swaps

$ /usr/bin/time deno run exprtk_functional_test.txt.js 
7445 7299
0.06user 0.02system 0:00.08elapsed 95%CPU (0avgtext+0avgdata 49888maxresident)k
0inputs+64outputs (0major+7174minor)pagefaults 0swaps
```

**Comment by mingodad** [comment]
And here is another non trivial program (a Javascript port of GLPK) to compare between quickjs, nodejs, bun, deno:
```
/usr/bin/time qjs --std mpl-all-19.js
52.95user 0.21system 0:53.18elapsed 99%CPU (0avgtext+0avgdata 380912maxresident)k
0inputs+0outputs (0major+96540minor)pagefaults 0swaps

/usr/bin/time node-env node mpl-all-19.js
4.43user 0.23system 0:03.25elapsed 143%CPU (0avgtext+0avgdata 286160maxresident)k
0inputs+0outputs (0major+66259minor)pagefaults 0swaps

/usr/bin/time bun mpl-all-19.js
3.85user 0.37system 0:03.12elapsed 135%CPU (0avgtext+0avgdata 292400maxresident)k
0inputs+0outputs (0major+72679minor)pagefaults 0swaps

/usr/bin/time deno run --allow-all mpl-all-19.js
4.46user 0.15system 0:02.91elapsed 158%CPU (0avgtext+0avgdata 191924maxresident)k
0inputs+64outputs (0major+46762minor)pagefaults 0swaps
```

[glpk-js.zip](https://github.com/oven-sh/bun/files/10346908/glpk-js.zip)

**Reply by mingodad** [reply]
After adding the missing functions to my build of https://github.com/mean-ui-thread/JavaScriptCore the memory usage is around 15% lower compared to `bun`:
```
/usr/bin/time myjsc mpl-all-19.js
4.10user 0.21system 0:03.17elapsed 135%CPU (0avgtext+0avgdata 253064maxresident)k
0inputs+0outputs (0major+63310minor)pagefaults 0swaps
```

**Reply by mingodad** [reply]
And here is the result for [hermes](https://hermesengine.dev):
```
/usr/bin/time node-hermes mpl-all-19.js
...
44.81user 0.10system 0:44.60elapsed 100%CPU (0avgtext+0avgdata 151252maxresident)k
0inputs+0outputs (0major+37818minor)pagefaults 0swaps
```
```
node-hermes --version
LLVM (http://llvm.org/):
  LLVH version 8.0.0svn
  Optimized build
```
